### PR TITLE
adding CCPP to the workflow

### PR DESCRIPTION
**Pull request by panll**
Modified files for CCPP option:
jobs/JFV3CAM_SAR_ENVIR : CCPP option added
parm/input_sar_conus_ccpp.nml : CCPP namelist
parm/suite_FV3_GFS_2017_gfdlmp_regional.xml : GFDL physics suite
scripts/exfv3cam_sar_fcst.sh: CCPP option added

It's tested on THEIA. Please test on other machines.
1) with CCPP
/scratch3/BMC/det/lpan/code_sprint/stmp/Linlin.Pan/tmpnwprd/forecast_tm00_conus_20170906_00
2)without CCPP
/scratch3/BMC/det/lpan/code_sprint/stmp/Linlin.Pan/tmpnwprd/forecast_tm00_conus_20170906_00_no_ccpp

### CODE REVIEW

File: jobs/JFV3CAM_SAR_ENVIR [hunk 1]
```diff
@@ -10,7 +10,7 @@ export tmpdir=stmp
 export comdir=ptmp
 export gesdir=ptmp
 export CCPP=false
-export CCPP_SUITE=FV3_GFS_2017_gfdlmp_regional
+expor tCCPP_SUITE=FV3_GFS_2017_gfdlmp_regional
```
Comment (JacobCarley-NOAA) [inline jobs/JFV3CAM_SAR_ENVIR]: Small type-o.  Change:

'export tCCPP_SUITE=FV3_GFS_2017_gfdlmp_regional'

 to 

'export CCPP_SUITE=FV3_GFS_2017_gfdlmp_regional

**Comment by panll** [comment]
Thanks!

I also added the build option to my fork, following files are modified:

jobs/JFV3CAM_SAR_ENVIR 
sorc/regional_build.cfg
sorc/partial_build.sh
sorc/build_forecast.sh
sorc/build_all.sh

**Comment by JacobCarley-NOAA** [comment]
@panll I just tried compiling and got the following error - is there something missing in the build step?:


> CCPP_layer/CCPP_data.F90(33): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [CCPP_TYPES]
>     use ccpp_types,    only: ccpp_t
> --------^
> CCPP_layer/CCPP_data.F90(70): error #6457: This derived type name has not been declared.   [CCPP_T]
>     type(ccpp_t),                                              save, target :: cdata_tile
> ---------^
> CCPP_layer/CCPP_data.F90(71): error #6457: This derived type name has not been declared.   [CCPP_T]
>     type(ccpp_t),                                              save, target :: cdata_domain
> ---------^
> CCPP_layer/CCPP_data.F90(72): error #6457: This derived type name has not been declared.   [CCPP_T]
>     type(ccpp_t),                 dimension(:,:), allocatable, save, target :: cdata_block
> ---------^
> CCPP_layer/CCPP_data.F90(33): error #6580: Name in only-list does not exist or is not accessible.   [CCPP_T]
>     use ccpp_types,    only: ccpp_t
> -----------------------------^
> CCPP_layer/CCPP_data.F90(43): error #6404: This name does not have a type, and must have an explicit type.   [CDATA_TILE]
>     public cdata_tile,             &
> -----------^
> CCPP_layer/CCPP_data.F90(44): error #6404: This name does not have a type, and must have an explicit type.   [CDATA_DOMAIN]
>            cdata_domain,           &
> -----------^
> CCPP_layer/CCPP_data.F90(45): error #6404: This name does not have a type, and must have an explicit type.   [CDATA_BLOCK]
>            cdata_block,            &
> -----------^
> compilation aborted for CCPP_layer/CCPP_data.F90 (code 1)
> gmake[2]: *** [CCPP_layer/CCPP_data.o] Error 1


Earlier in my log file I also found this:


> ld: final link failed: Nonrepresentable section on output
> make[3]: *** [physics/libccppphys.so.3.0.0] Error 1
> make[3]: Target `physics/CMakeFiles/ccppphys.dir/build' not remade because of errors.
> make[3]: Leaving directory `/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/ccpp/build'
> make[2]: *** [physics/CMakeFiles/ccppphys.dir/all] Error 2
> make[2]: Target `all' not remade because of errors.
> make[2]: Leaving directory `/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/ccpp/build'
> make[1]: *** [all] Error 2
> make[1]: Target `default_target' not remade because of errors.
> make[1]: Leaving directory `/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/ccpp/build'
> gmake: *** [/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/ccpp/ccpp.mk] Error 2
>

**Comment by panll** [comment]
@JacobCarley-NOAA The error message suggests missing CCPP code,  is the code the latest code from vlab master? could you send me your whole compiling log file? I'll check it. Thanks!

**Comment by JacobCarley-NOAA** [comment]
> @JacobCarley-NOAA The error message suggests missing CCPP code, is the code the latest code from vlab master? could you send me your whole compiling log file? I'll check it. Thanks!

I just took a look at the Externals.cfg file and it appears that it's pointing to the 'regional' branch of the NEMSfv3gfs repo.  But it looks like @BinLiu-NOAA merged the master into this on 7/17

https://vlab.ncep.noaa.gov/redmine/projects/nemsfv3gfs/repository?utf8=%E2%9C%93&rev=regional

I've also attached the compile log in case there is anything helpful in there.

[build_forecast_ccpp.log](https://github.com/NOAA-EMC/regional_workflow/files/3433028/build_forecast_ccpp.log)

**Comment by climbfuji** [comment]
@JacobCarley-NOAA @panll I am chiming in here, because I was working with Jun and Dusan to port the CCPP code to wcoss_dell_p3 and wcoss_cray. The issue is that you need to use a different version of the NCEPlibs. The version you are pointing to (operational NCO version) has not been compiled with '-fPIC' (to generate position independent code) and is also not thread-safe (i.e. not compiled with the OpenMP flags).

Are you using the static or dynamic CCPP?

**Comment by panll** [comment]
it's dynamic build @JacobCarley-NOAA @climbfuji

**Comment by climbfuji** [comment]
You should be able to get away by using the static build, but on the long run we need to have the thread-safe and fPIC-ed NCEPlibs on the WCOSS machines. Tagging a few people here for their awareness.

@JacobCarley-NOAA @panll @DusanJovic-NOAA @junwang-noaa

**Comment by JacobCarley-NOAA** [comment]
Thanks @climbfuji 

To perform the static build do I simply pass "STATIC=Y" as well as a specific a suite?

**Comment by climbfuji** [comment]
Yes, please add STATIC=Y SUITES=XYZ where the suite name XYZ corresponds to an existing suite definition file name with path/name ccpp/suites/suite_XYZ.xml. You can specify multiple suites for one build with SUITES=XYZ,ABC,...

**Comment by JacobCarley-NOAA** [comment]
@climbfuji Thanks for your help!

Apologies if I am missing something obvious here (and I suspect I am) but I've tried specifying a suite and get the following error:

> gmake -j 8 -k COMPONENTS=CCPP,FMS,FV3 TEST_BUILD_NAME=fv3_1 BUILD_ENV=wcoss_dell_p3 'FV3_MAKEOPT=NCEP64LEV=Y HYDRO=N 32BIT=Y CCPP=Y STATIC=Y SUITE=/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/ccpp/suites/suite_FV3_GFS_2017_gfdlmp_regional.xml' distclean
> Will copy modules.nems and NEMS.x as fv3_1 under tests/
> NOTE: Skipping appbuilder.mk creation; no appbuilder file in use.
> echo 'FMS CCPP FV3' > "/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/NEMS/src/conf/components.mk"
> Adding FV3 makeopts to FMS makeopts
> /gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/NEMS/src/incmake/component_CCPP.mk:33: *** Option STATIC=Y requires suites argument as SUITES=xyz,abc,... (where suite xyz corresponds to file suite_xyz.xml).  Stop.

I've attached the full extent of the log file as well in case that's helpful.  Note I also checked the suite definition file does exist at the path noted above.
[build_forecast_ccpp.log](https://github.com/NOAA-EMC/regional_workflow/files/3436879/build_forecast_ccpp.log)

**Comment by llpcarson** [comment]
Jacob - The argument to compile.sh is *only* the name of the suite, not the full file name, i.e.

SUITE=FV3_GFS_2017_gfdlmp_regional

The full file name is constructed from this...  give that a try!

**Comment by JacobCarley-NOAA** [comment]
Thanks for the suggestion @llpcarson - unfortunately I was met with the same error when I gave that a try,  Last bit of the log file follows:

> gmake -j 8 -k COMPONENTS=CCPP,FMS,FV3 TEST_BUILD_NAME=fv3_1 BUILD_ENV=wcoss_dell_p3 'FV3_MAKEOPT=NCEP64LEV=Y HYDRO=N 32BIT=Y CCPP=Y STATIC=Y SUITE=FV3_GFS_2017_gfdlmp_regional' distclean
> Will copy modules.nems and NEMS.x as fv3_1 under tests/
> NOTE: Skipping appbuilder.mk creation; no appbuilder file in use.
> echo 'FMS CCPP FV3' > "/gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/NEMS/src/conf/components.mk"
> Adding FV3 makeopts to FMS makeopts
> /gpfs/dell2/emc/modeling/noscrub/Jacob.Carley/sprint/ccpp/regional_workflow/sorc/regional_forecast.fd/NEMS/src/incmake/component_CCPP.mk:33: *** Option STATIC=Y requires suites argument as SUITES=xyz,abc,... (where suite xyz corresponds to file suite_xyz.xml).  Stop.

**Comment by llpcarson** [comment]
Ah - you got caught between documentation and updated features!  Use SUITES=xyz instead of SUITE=xyz (even if you only have one)...  part of the "build multiple suites" update!

Laurie

**Comment by JacobCarley-NOAA** [comment]
@llpcarson Aha!  That was it - it's compiling now.  Thanks a bunch Laurie!

**Comment by BenjaminBlake-NOAA** [comment]
Hi Linlin, I got the CCPP code to compile on the WCOSS Dell and I am waiting for my test forecast to run.  I had to wait several hours yesterday for my job to run, only for it to fail immediately because of an incorrect path.  Hopefully I will be able to complete this soon!

**Comment by panll** [comment]
Thanks for the updates! Please let me know if you need any help! @BenjaminBlake-NOAA

**Comment by BenjaminBlake-NOAA** [comment]
Thanks Linlin!  Unfortunately I wasn't able to check the output of the test I submitted yesterday because WCOSS went down for maintenance.  It will be returned on Friday.  Sorry it has taken this long to run the test.

**Comment by BenjaminBlake-NOAA** [comment]
Hi Linlin, I apologize this has taken so long for me to test but I was able to get a successful forecast to run on WCOSS Dell with the CCPP forecast executable.  After I run a post job and confirm the output values look fine I'll merge the pull request.

**Comment by panll** [comment]
Hi, Ben, That's great to know the CCPP ran on WCOSS machine successfully. Thanks a lot for your efforts!
@BenjaminBlake-NOAA

**Comment by BenjaminBlake-NOAA** [comment]
Happy to help!  My post job completed successfully so I will merge the pull request now.

**Pull request merged**